src/zkp/ecdsa_adaptor.rs

```diff
@@ -0,0 +1,317 @@
+//! # ECDSA Adaptor
+//! Support for ECDSA based adaptor signatures.
+//!
+
+use core::{fmt, ptr, str};
+use ffi::{self, CPtr, ECDSA_ADAPTOR_SIGNATURE_LENGTH};
+use {constants, PublicKey, Secp256k1, SecretKey};
+use {from_hex, Error};
+use {Message, Signing};
+use {Signature, Verification};
+
+/// Represents an adaptor signature and dleq proof.
+#[derive(Debug, PartialEq, Clone, Copy, Eq)]
+pub struct EcdsaAdaptorSignature(ffi::EcdsaAdaptorSignature);
+
+impl fmt::LowerHex for EcdsaAdaptorSignature {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        for ch in self.0.as_bytes().iter() {
+            write!(f, "{:02x}", ch)?;
+        }
+        Ok(())
+    }
+}
+
+impl fmt::Display for EcdsaAdaptorSignature {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        fmt::LowerHex::fmt(self, f)
+    }
+}
+
+impl str::FromStr for EcdsaAdaptorSignature {
+    type Err = Error;
+    fn from_str(s: &str) -> Result<EcdsaAdaptorSignature, Error> {
+        let mut res = [0; ECDSA_ADAPTOR_SIGNATURE_LENGTH];
+        match from_hex(s, &mut res) {
+            Ok(ECDSA_ADAPTOR_SIGNATURE_LENGTH) => {
+                EcdsaAdaptorSignature::from_slice(&res[0..ECDSA_ADAPTOR_SIGNATURE_LENGTH])
+            }
+            _ => Err(Error::InvalidEcdsaAdaptorSignature),
+        }
+    }
+}
+
+impl CPtr for EcdsaAdaptorSignature {
+    type Target = ffi::EcdsaAdaptorSignature;
+    fn as_c_ptr(&self) -> *const Self::Target {
+        self.as_ptr()
+    }
+
+    fn as_mut_c_ptr(&mut self) -> *mut Self::Target {
+        self.as_mut_ptr()
+    }
+}
+
+impl EcdsaAdaptorSignature {
+    /// Creates an AdaptorSignature directly from a slice
```

Comment:
The spelling of "AdaptorSignature" here is inconsistent with the rest of the file.

Reply:
Thanks fixed (I assumed you meant it should be `EcdsaAdadptorSignature`?)

Reply:
Sorry I should have been clearer!

In the rest of the documentation, it is usually spelt "adaptor signature" and only here we use a PascalCase for some reason. Given that we are not referring to any type, I think the normal English spelling is preferable.

Reply:
If you wanted to explicitly reference the type, you can use the square-brackets + backtick notation:

```rust
/// Creates an [`EcdsaAdaptorSignature`] directly from a slice
```

That will render to an actual link in rustdoc and on a recent toolchain, it will emit a warning if the type you are referencing doesn't actually exist!

Reply:
Thanks that's neat, I changed to using that. I think it makes sense to reference the type here because it's somehow different than the `encrypt` function in the sense that the goal is to create the struct directly and not generate it. I don't have a strong opinion though so I can change it to plain English if you prefer.